[Bryann Yeap] iP

BryannYeap opened the pull request:
# DUKE (INDIVIDUAL PROJECT)

## Gary Bot 🤖 

> “There is satisfaction," he said to Dalinar, "in creating a list of things you can actually accomplish, 
> then removing them one at a time.” - *Brandon Sanderson, author of Oathbringer* ([source](https://www.goodreads.com/quotes/10211374-there-is-satisfaction-he-said-to-dalinar-in-creating-a))


Gary makes sure you keep tabs on all the things that you have to do! Gary would be **very** disappointed if you forget to finish your tasks... ~~You don't want to know the things he'd do...~~


Here are some attributes of Gary:
* Text-based CLI
* Intuitive and simple
* Responsive


To get Gary, just follow these simple steps:
1. Download from [here](https://github.com/BryannYeap/ip/releases).
2. Open your command line and navigate to the directory containing the `iP.jar` file.
3. Type this command: `java -jar {filename}.jar` where `{filename}` is the name of the file (duh!)
4. Add your tasks and let Gary take care of the rest!  


Features:
- [x] Basic functionality of adding, removing and marking tasks as done.
- [x] Querying / Searching for tasks.
- [x] Built as a downloadable jar
- [x] GUI (coming soon)
- [ ] Adding times to Deadline tasks


It is a (hopefully) elegant program as well! Here is the `main` method:
```java
    public static void main(String[] args) {
        new Duke().run();
    }
```

Commit:
Add Gradle support

BryannYeap committed:
Add Level-1

BryannYeap committed:
Add Level-2

BryannYeap committed:
Add Level-3

BryannYeap committed:
Add Level-4

BryannYeap committed:
Add A-TextUiTesting

BryannYeap committed:
Add Level-5

BryannYeap committed:
Add Level-6

BryannYeap committed:
Add A-Enums

BryannYeap committed:
Refactor code

BryannYeap committed:
Fix coding style to follow reccomended coding standards

BryannYeap committed:
Add Level-7

BryannYeap committed:
Change switch block cases from string to char

BryannYeap committed:
Add Level-8

BryannYeap committed:
Merge branch 'branch-Level-7'

BryannYeap committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java

BryannYeap committed:
Add TaskList Class

BryannYeap committed:
Add A-MoreOOP

BryannYeap committed:
Add A-Packages

BryannYeap committed:
Add three sub-packages

BryannYeap committed:
Add A-JUnit with 5 JUnit test classes

BryannYeap committed:
Add A-Jar

BryannYeap committed:
Refactor code by changing constructors and shifting methods

BryannYeap committed:
Add more JUnit tests for DeadlineTest and EventTest

BryannYeap committed:
Add A-JavaDoc

BryannYeap committed:
Add A-CodingStandard

BryannYeap committed:
Add Level-9

BryannYeap committed:
Merge branch 'branch-A-CodingStandard'

BryannYeap committed:
Merge branch 'branch-Level-9'

BryannYeap committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/main/TaskList.java

BryannYeap committed:
Refactor code by shifting methods and import orders

BryannYeap committed:
Merge branch 'add-gradle-support'

HolmesJJ commented on src/test/java/duke/task/TodoTest.java:
```diff
@@ -0,0 +1,19 @@
+
+public class TodoTest {
+     @Test
+     public void toString_descriptionOfTodo1_todoStringReturned() {
```
Perhaps it would be better to merge these two test cases in the same method and add message parameter in `assertEquals` method to distinguish them rather than name it as `toString_descriptionOfTodo1` and `toString_descriptionOfTodo2` since I think they are testing the same method in the same scenario and returning the same type of result, and the naming of 1 and 2 looks confusing? 🤔

BryannYeap replied:
Done!

HolmesJJ commented on src/test/java/duke/task/TaskTest.java:
```diff
@@ -0,0 +1,38 @@
+
+import org.junit.jupiter.api.Test;
+
+public class TaskTest {
```
I like your very complete test cases! 😄

BryannYeap replied:
Thanks!

HolmesJJ commented on src/test/java/duke/task/EventTest.java:
```diff
@@ -0,0 +1,41 @@
+
+import duke.exception.InvalidParamException;
+
+public class EventTest {
```
I like your very complete test cases +1! 😄

HolmesJJ commented on src/test/java/duke/task/DeadlineTest.java:
```diff
@@ -0,0 +1,50 @@
+
+import duke.exception.InvalidParamException;
+
+public class DeadlineTest {
```
I like your very complete test cases +1! 😄

HolmesJJ commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,208 @@
+    private void findTask(String input) {
+        try {
+            System.out.println("Output: These tasks have descriptions that contain the phrase '"
+                    + input.substring(5) + "'!\n");
```
Should this magic number be moved to top by using named constant? 🤔

BryannYeap replied:
Done!

HolmesJJ commented on src/main/java/duke/task/Storage.java:
```diff
@@ -0,0 +1,122 @@
+                toBeWritten.append(initialOfTask).append(SEPARATOR).append(description);
+
+                switch (initialOfTask) {
+                case 'T':
```
```
private static final char TASK_TYPE = 'T';
```
Should this magic character be moved to top by using named constant or use enum? 🤔
And I noticed the same issue in several other places below.

BryannYeap replied:
Hey I tried taking this into consideration and changing it to an enum but apparently it doesn't work. Switch block cases require 'constant expressions', and even when I set the enum field of `initial` to be `final`, it did not work. So I guess I'll just leave it as it is. Thanks for the suggestion!

HolmesJJ commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,208 @@
+        try {
+            if (!input.startsWith("todo") && !input.startsWith("deadline") && !input.startsWith("event")) {
+                throw new duke.exception.InvalidInputException();
+            } else {
```
Perhaps the `else` can be removed ? 🤔

BryannYeap replied:
Done!

HolmesJJ reviewed (commented):
LGTM! I like your super clear code and all the code structure is very complete. Only a few code suggestions and magic numbers and characters should be moved to top by using named constant or use enum. Well done! 😃

Agentum07 commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,208 @@
+    private void findTask(String input) {
+        try {
+            System.out.println("Output: These tasks have descriptions that contain the phrase '"
+                    + input.substring(5) + "'!\n");
```
I could not understand the purpose of the '5'. I suggest you declare it as a constant so that it's easier to understand the context?

BryannYeap replied:
Done!

Agentum07 commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,208 @@
+        try {
+            if (!input.startsWith("todo") && !input.startsWith("deadline") && !input.startsWith("event")) {
+                throw new duke.exception.InvalidInputException();
+            } else {
```
I think the `else` is not needed.

Agentum07 commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,208 @@
+        try {
+            if (!input.startsWith("todo") && !input.startsWith("deadline") && !input.startsWith("event")) {
+                throw new duke.exception.InvalidInputException();
+            } else {
```
I think you can remove this `else`. If the `if` condition is satisfied, the program will directly go to the catch and skip all the code

BryannYeap replied:
Done!

Agentum07 reviewed (commented):
Overall, pretty well written code. Quite a good job with the javadocs, very easy to understand! Good job!

BryannYeap committed:
Fix code based on PR reviews

BryannYeap committed:
Add A-Checkstyle

BryannYeap committed:
Merge branch 'branch-A-Checkstyle'

MGRL2201 reviewed (commented):
Very clean, well written, and well-documented code!

Chesterwongz reviewed (commented):
LGTM! Very nice and clean code.

BryannYeap committed:
Add Level-10

BryannYeap committed:
Merge branch 'branch-Level-10'

BryannYeap committed:
Fix checkstyle and add javadocs for new classes and methods

BryannYeap committed:
Refactor code to shift methods and change some javadocs

BryannYeap committed:
Add assertions.

There are comments signifying certain invariants, as well as default
cases in switch blocks that have an empty body.

Comments do not functionally stop the code if the commented invariant is
violated. Empty default cases in switch blocks also work under
assumptions that are not watched out for by the compiler.

Let's substitute comments with assertions, as well as add assertions
into the empty default case bodies.

Using assertions is preferable over comments to result in the program
telling us if invariants were to be violated.

BryannYeap committed:
Add assertions.

There are comments signifying certain invariants, as well as default
cases in switch blocks that have an empty body.

Comments do not functionally stop the code if the commented invariant is
violated. Empty default cases in switch blocks also work under
assumptions that are not watched out for by the compiler.

Let's substitute comments with assertions, as well as add assertions
into the empty default case bodies.

Using assertions is preferable over comments to result in the program
telling us if invariants were to be violated.

BryannYeap committed:
Improve code quality

Code violates many code quality practices, such as inappropriately
scopes variables, code that can be refactored to omit explanatory
comments, magic literals, long methods and SLAP violations.

Let's refactor the code in order to remove these violations and improve
code quality.

Doing so is important to improve readability and results in cleaner
code.

BryannYeap committed:
Improve code quality

Code violates many code quality practices, such as inappropriatesly
scopes variables, code that can be refactored to omit explanatory
comments, magic literals, long methods and SLAP violations.

Let's refactor the code in order to remove these violations and improve
code quality.

Doing so is important to improve readability and results in cleaner
code.

BryannYeap committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/BryannYeap/ip into branch-A-CodeQuality

BryannYeap committed:
Merge branch 'master' of https://github.com/BryannYeap/ip

BryannYeap committed:
Merge branch 'master' into branch-A-Assertions

BryannYeap committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/duke/main/Parser.java

BryannYeap committed:
Merge branch 'branch-A-Assertions' of https://github.com/BryannYeap/ip into branch-A-Assertions

# Conflicts:
#	src/main/java/duke/main/Parser.java

BryannYeap committed:
Merge pull request #2 from BryannYeap/branch-A-CodeQuality

Add  A-CodeQuality

BryannYeap committed:
Merge branch 'master' into branch-A-CodeQuality

BryannYeap committed:
Merge branch 'master' into branch-A-Assertions

BryannYeap committed:
Merge pull request #1 from BryannYeap/branch-A-Assertions

Add A-Assertions.

BryannYeap committed:
Merge branch 'master' into branch-A-Assertions

BryannYeap committed:
Fix line that conflict broke

BryannYeap committed:
Add A-Streams

BryannYeap committed:
Merge pull request #3 from BryannYeap/branch-A-Streams

Add A-Streams

BryannYeap committed:
Add BCD-Extension by adding C-Update

There is no update function to update a specific Task in the TaskList.

Considering that the TaskList cannot be sorted or ordered, the user
potentially wants to manually add Tasks in a certain order. However, if
the user wanted to change a certain Task, he would have to delete that
Task, as well as all following Tasks, to manually sort them.

Let's add an update function so that a specific Task at a given index
be changed or modified without having to delete and then add a new Task.

Incorporating an update function would result in increased flexibility
in the functionality of the application and how it can be used.

BryannYeap committed:
Merge branch 'branch-A-Assertions'

BryannYeap committed:
Merge branch 'branch-A-CodeQuality'

BryannYeap committed:
Add A-CI

BryannYeap committed:
Remove A-CI

BryannYeap committed:
Revert "Add A-CI"

This reverts commit e59e2c6547f8d9f0387634abb36ed0774e752231.

BryannYeap committed:
Revert "Remove A-CI"

This reverts commit 279b66a23490e3c4f457d595f3d0164c42da1c15.

BryannYeap committed:
Revert "Revert "Remove A-CI""

This reverts commit 4d985e6af4645ea4204abad51b6f6292ea6d9bd6.

BryannYeap committed:
Add A-CI

BryannYeap committed:
Fix checkstyle errors

BryannYeap committed:
Merge pull request #4 from BryannYeap/BryannYeap-patch-1

Add A-CI

BryannYeap committed:
Increase readability and quality of code

BryannYeap committed:
Merge branch 'master' of https://github.com/BryannYeap/ip

BryannYeap committed:
Add user guide

BryannYeap committed:
Set theme jekyll-theme-dinky

BryannYeap committed:
Delete theme

BryannYeap committed:
Fix user guide GFMD

BryannYeap committed:
Fix user guide GFMD

BryannYeap committed:
Fix more user guide GFMD

BryannYeap committed:
Fix user guide website

BryannYeap committed:
Fix emoji and add inline code to some sentences

BryannYeap committed:
Fix user guide website command summary table

BryannYeap committed:
Fix user guide Gary image

BryannYeap committed:
Add exit action to command summary table

BryannYeap committed:
Fix GUI truncation and other UI elements

BryannYeap committed:
Change user guide GUI example picture
